### PR TITLE
PUBDEV-6054 - Fixed unclosed leaked socket in RabitWorker

### PR DESCRIPTION
**Pull request by Pscheidl**
https://0xdata.atlassian.net/browse/PUBDEV-6054

Connections are not being closed, as pointer to tracker thread is being deleted prematurely. The resources were garbage collected, but sometimes not quickly enough. In case of building many small models at once, the resources were never freed in time, resulting in SocketChannels holding file descriptors allocated on Linux.

### CODE REVIEW

**Comment by Pscheidl** [comment]
An attempt to make the test quicker using Ulimit.

```
The latest docker supports setting ulimits through the command line and the API. For instance, docker run takes --ulimit <type>=<soft>:<hard> and there can be as many of these as you like. So, for your nofile, an example would be --ulimit nofile=262144:262144
```

**Comment by Pscheidl** [comment]
Moved the stress test to the smoke stage (thanks @michal-raska ) to see the output fast: http://mr-0xc1:8080/job/h2o-3-pipeline/job/PR-3051/9/

Now moved to nigthly. The test itself is not that slow, finishes in +- 5 minutes. Yet I'd leave it in nightly.

File: h2o-extensions/xgboost/src/main/java/hex/tree/xgboost/rabit/RabitTrackerH2O.java [hunk 1]
```diff
@@ -71,9 +71,8 @@ public boolean start(long timeout) {
             throw new IllegalStateException("Rabit tracker already started.");
         }
         RabitTrackerH2OThread trackerThread = new RabitTrackerH2OThread(this);
-        trackerThread.setDaemon(true);
```
Comment (Pscheidl) [inline h2o-extensions/xgboost/src/main/java/hex/tree/xgboost/rabit/RabitTrackerH2O.java]: There is no reason for this thread to be a daemon thread. It should always be killed in time. And if JVM with H2O is being stopped, the thread should be stopped too.

File: h2o-extensions/xgboost/src/main/java/hex/tree/xgboost/rabit/RabitTrackerH2O.java [hunk 2]
```diff
@@ -71,9 +71,8 @@ public boolean start(long timeout) {
...
-        trackerThread.start();
         this.trackerThread = trackerThread;
+        trackerThread.start();
```
Comment (Pscheidl) [inline h2o-extensions/xgboost/src/main/java/hex/tree/xgboost/rabit/RabitTrackerH2O.java]: First assign it, then start it. If start() throws an unechecked exception, the pointer is still there in `this.trackerThread` and can be dealth with later.

File: h2o-extensions/xgboost/src/main/java/hex/tree/xgboost/rabit/RabitTrackerH2O.java [hunk 3]
```diff
@@ -20,7 +20,7 @@
 
     private Map<String, String> envs = new HashMap<>();
 
-    private RabitTrackerH2OThread trackerThread;
+    private volatile RabitTrackerH2OThread trackerThread;
```
Comment (Pscheidl) [inline h2o-extensions/xgboost/src/main/java/hex/tree/xgboost/rabit/RabitTrackerH2O.java]: Red from both the thread and this class. Better guarantee single place of existence.

File: h2o-extensions/xgboost/src/main/java/hex/tree/xgboost/rabit/RabitTrackerH2O.java [hunk 4]
```diff
@@ -71,26 +71,28 @@ public boolean start(long timeout) {
...
         this.trackerThread = trackerThread;
+        trackerThread.start();
         return true;
     }
 
     @Override
     public void stop() {
-        if(null != this.trackerThread) {
-            this.trackerThread.interrupt();
+        assert this.trackerThread != null;
```
Comment (Pscheidl) [inline h2o-extensions/xgboost/src/main/java/hex/tree/xgboost/rabit/RabitTrackerH2O.java]: Null check removed. This is never assigned null before this method's invocation now.

In fact, it should never ever be null at this point. That's why there is this new assert.

File: h2o-extensions/xgboost/src/main/java/hex/tree/xgboost/rabit/RabitTrackerH2O.java [hunk 5]
```diff
@@ -71,26 +71,28 @@ public boolean start(long timeout) {
...
-                this.sock.close();
-            } catch (IOException e) {
-                throw new RuntimeException("Failed to close Rabit tracker socket.", e);
-            } finally {
-                this.port = 9091;
-            }
+
+        try {
+            this.sock.close();
+            this.port = 9091;
```
Comment (Pscheidl) [inline h2o-extensions/xgboost/src/main/java/hex/tree/xgboost/rabit/RabitTrackerH2O.java]: Only reset the port to default when socket closing is successful. 

When unsuccessful, log an error and continue, as there may be enough resources to finish the task. The error will be in the logs still.

File: scripts/jenkins/groovy/defineTestStages.groovy [hunk 6]
```diff
@@ -311,6 +311,10 @@ def call(final pipelineContext) {
     [
       stageName: 'R3.3 CMD Check as CRAN', target: 'test-r-cmd-check-as-cran', rVersion: '3.3.3',
       timeoutValue: 10, hasJUnit: false, component: pipelineContext.getBuildConfig().COMPONENT_R
+    ],
+    [ // These run with reduced number of file descriptors for early detection of FD leaks
+      stageName: 'XGBoost Stress tests', target: 'test-pyunit-xgboost-stress', pythonVersion: '3.5', timeoutValue: 20,
+      component: pipelineContext.getBuildConfig().COMPONENT_PY, customDockerArgs: [ '--ulimit nofile=100:100' ]
```
Comment (Pscheidl) [inline scripts/jenkins/groovy/defineTestStages.groovy]: Notice the `'--ulimit nofile=100:100'`. Makes this test fail fast if there is something wrong with resources not being freed. If these tests are ran without the fixes, they fail in less than 100 models built (as some of the descriptors are already taken by the system).

File: h2o-py/tests/testdir_algos/xgboost/stress/pyunit_xgboost_airlines_stress.py [hunk 7]
```diff
@@ -0,0 +1,18 @@
+import h2o
```
Comment (michalkurka) [inline h2o-py/tests/testdir_algos/xgboost/stress/pyunit_xgboost_airlines_stress.py]: should we mark this test as large?

Comment (michalkurka) [reply h2o-py/tests/testdir_algos/xgboost/stress/pyunit_xgboost_airlines_stress.py]: sorry, I didn't notice there was already a special category for it

File: scripts/jenkins/Makefile.jenkins [hunk 8]
```diff
@@ -29,7 +29,7 @@ test-package-py:
 	zip -q -r test-package-py.zip h2o-py/tests/ h2o-py/demos/ \
 	h2o-docs/src/booklets/v2_2015/source h2o-py/build/dist/*.whl \
 	h2o-genmodel/build/libs/h2o-genmodel.jar h2o-assemblies/genmodel/build/libs/genmodel.jar scripts/run.py \
-	$(JENKINS_MAKEFILE_PATH) tests/pyunitSmokeTestList tests/pyunitAutoMLList tests/pyunitSingleNodeList h2o-py/scripts/h2o-py-test-setup.py
```
Comment (michalkurka) [inline scripts/jenkins/Makefile.jenkins]: @michal-raska can you please check this?

File: scripts/jenkins/Makefile.jenkins [hunk 9]
```diff
@@ -59,6 +59,10 @@ test-pyunit-single-node:
 	cd h2o-py/tests/ && rm -f -r testdir_hdfs
 	export NO_GCE_CHECK=True && cd h2o-py/tests/ && ../../scripts/run.py --wipeall --geterrs --testsize s --numclouds 1 --numnodes 1 --jvm.xmx 3g --testlist ../../tests/pyunitSingleNodeList
 
+test-pyunit-xgboost-stress:
```
Comment (michalkurka) [inline scripts/jenkins/Makefile.jenkins]: great!

**Comment by michalkurka** [comment]
great work! thank you

approved from my side - please let @michal-raska review the build changes

**Comment by Pscheidl** [comment]
@pkozelka also reviewed, gave thumbs up :)

**Pull request merged**